Source/GSThemeDrawing.m

```diff
@@ -3475,15 +3496,343 @@ - (void) drawTableViewRow: (NSInteger)rowIndex
+              drawingRect.origin.x
+                += indentationFactor + imageRect.size.width + 5;
+              drawingRect.size.width
+                -= indentationFactor + imageRect.size.width + 5;
+            }
+          else
+            {
+              drawingRect.origin.x += indentationFactor;
+              drawingRect.size.width -= indentationFactor;
+            }
+
+          RELEASE(imageCell);
+        }
+
+      [cell drawWithFrame: drawingRect inView: outlineView];
+      if (i == editedColumn && rowIndex == editedRow)
+        {
+          [cell _setInEditing: NO];
+          [cell setShowsFirstResponder: NO];
+        }
+    }  
+}
+
+- (void) drawCellViewRow: (NSInteger)rowIndex
+		clipRect: (NSRect)clipRect
+		  inView: (id)v
+{
+  NSInteger numberOfColumns = [v numberOfColumns];
+  CGFloat *columnOrigins = [v _columnOrigins];
+  NSArray *tableColumns = [v tableColumns];
+  NSInteger numberOfRows = [v numberOfRows];
+  NSInteger startingColumn; 
+  NSInteger endingColumn;
+  NSInteger i;
+  CGFloat x_pos;
+  id dataSource = [v dataSource];
+  id delegate = [v delegate];
+  BOOL hasMethod = NO;
+  NSTableColumn *outlineTableColumn = nil;
+
+  // If we have no data source, there is nothing to do...
+  if (dataSource == nil)
+    {
+      return;
+    }
+
+  // If the rowIndex is greater than the numberOfRows, done...
+  if (rowIndex >= numberOfRows)
+    {
+      return;
+    }
+  
+  // Check the delegate method...
+  if ([v isKindOfClass: [NSOutlineView class]])
+    {
+      outlineTableColumn = [v outlineTableColumn];
+      hasMethod = [delegate respondsToSelector: @selector(outlineView:viewForTableColumn:item:)];
+    }
+  else if ([v isKindOfClass: [NSTableView class]])
+    {
+      hasMethod = [delegate respondsToSelector: @selector(tableView:viewForTableColumn:row:)];
+    }
+  
+  /* Determine starting column as fast as possible */
+  x_pos = NSMinX (clipRect);
+  i = 0;
+  while ((i < numberOfColumns) && (x_pos > columnOrigins[i]))
+    {
+      i++;
+    }
+  startingColumn = (i - 1);
+
+  if (startingColumn == -1)
+    startingColumn = 0;
+
+  /* Determine ending column as fast as possible */
+  x_pos = NSMaxX (clipRect);
+  while ((i < numberOfColumns) && (x_pos > columnOrigins[i]))
+    {
+      i++;
+    }
+  endingColumn = (i - 1);
+
+  if (endingColumn == -1)
+    endingColumn = numberOfColumns - 1;
+
+  /* Draw the row between startingColumn and endingColumn */
+  for (i = startingColumn; i <= endingColumn; i++)
+    {
+      NSRect drawingRect = [v frameOfCellAtColumn: i
+					      row: rowIndex];
+      NSTableColumn *tb = [tableColumns objectAtIndex: i];
+      NSIndexPath *path = [NSIndexPath indexPathForItem: i
+					      inSection: rowIndex];
+      NSMapTable *paths = [v _renderedViewPaths];
+      NSMapTable *views = [v _pathsToViews];
+      NSView *view = [paths objectForKey: path];
+
+      if ([v isKindOfClass: [NSOutlineView class]])
```

Comment:
Again having separate helper methods on the classes themselves would remove this ugly code.